### PR TITLE
SPV Bitcoin HTLC

### PR DESCRIPTION
**Pull request by Bushstar**
Adds the following RPC calls:

spv_getaddresspubkey - Returns the pubkey for the provided address.
spv_createhtlc - Create a HTLC script
spv_claimhtlc - Claim outputs in a HTLC script
spv_refundhtlc - Refund outputs in a HTLC script
spv_listhtlcoutputs - List all outputs in a HTLC script
spv_decodehtlcscript - Decode and HTLC script
spv_gethtlcseed - Get seed from HTLC claim TX

More information on these calls can be found here:
https://github.com/DeFiCh/ain/wiki/SPV-RPC-Calls

### CODE REVIEW

File: src/spv/bitcoin/BRTransaction.cpp [hunk 1]
```diff
@@ -651,17 +673,27 @@ int BRTransactionSign(BRTransaction *tx, int forkId, BRKey keys[], size_t keysCo
...
+        const uint8_t *hash;
+
+        if (htlcType == ScriptTypeNone)
+        {
+            hash = BRScriptPKH(input->script, input->scriptLen);
+        }
+        else
+        {
+            UInt160 hash160 = BRHTLCScriptPKH(input->script, input->scriptLen, htlcType);
+            hash = &hash160.u8[0];
```
Comment (bvbfan) [inline src/spv/bitcoin/BRTransaction.cpp:685]: Pointer to temporary address

File: src/spv/spv_wrapper.cpp [hunk 2]
```diff
@@ -662,13 +700,218 @@ UniValue CSpvWrapper::ListTransactions()
...
+uint64_t CSpvWrapper::GetFeeRate()
+{
+    return BRWalletFeePerKb(wallet);
+}
+
+HTLCDetails GetHTLCDetails(CScript& redeemScript)
+{
+    HTLCDetails script;
+
+    if (redeemScript[2] != 32)
```
Comment (bvbfan) [inline src/spv/spv_wrapper.cpp:779]: Can we verify that we have 32 bytes in redeemscript

File: src/spv/spv_wrapper.cpp [hunk 3]
```diff
@@ -662,13 +700,218 @@ UniValue CSpvWrapper::ListTransactions()
...
+    {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Incorrect seed hash length");
+    }
+    script.hash = std::vector<unsigned char>(&redeemScript[3], &redeemScript[35]);
+
+    uint8_t sellerLength = redeemScript[36];
```
Comment (bvbfan) [inline src/spv/spv_wrapper.cpp:785]: Here as well, [35] is size of following data, let's check it really presents

**Review by bvbfan — CHANGES_REQUESTED** [review]

File: src/spv/bitcoin/BRTransaction.cpp (hunk 1 above)
Comment (Bushstar) [reply src/spv/bitcoin/BRTransaction.cpp:686]: Good spot.

File: src/spv/spv_wrapper.cpp (hunk 2 above)
Comment (Bushstar) [reply src/spv/spv_wrapper.cpp:787]: Now building up script length as we parse it and check it at each step.

File: src/spv/spv_wrapper.cpp (hunk 3 above)
Comment (Bushstar) [reply src/spv/spv_wrapper.cpp:793]: Resolved in https://github.com/DeFiCh/ain/pull/308/commits/6af59ba1a8392ffb54b71bed8b51c78d54913dec

File: src/spv/bitcoin/BRWallet.cpp [hunk 4]
```diff
@@ -829,7 +960,15 @@ BRTransaction *BRWalletCreateTxForOutputs(BRWallet *wallet, const BRTxOutput out
...
-    feeAmount = _txFee(wallet->feePerKb, BRTransactionVSize(transaction) + TX_OUTPUT_SIZE);
+    if (feeRate > 1000)
+    {
+        feeAmount = _txFee(feeRate, BRTransactionVSize(transaction) + TX_OUTPUT_SIZE);
+    }
+    else
+    {
+        feeAmount = _txFee(wallet->feePerKb, BRTransactionVSize(transaction) + TX_OUTPUT_SIZE);
+    }
+
```
Comment (bvbfan) [inline src/spv/bitcoin/BRWallet.cpp:971]: We can do
```
feeRate = std::max(feeRate, wallet->feePerKb);
feeAmount = _txFee(feeRate, BRTransactionVSize(transaction) + TX_OUTPUT_SIZE);
```

File: src/spv/bitcoin/BRWallet.cpp [hunk 5]
```diff
@@ -875,7 +1020,14 @@ BRTransaction *BRWalletCreateTxForOutputs(BRWallet *wallet, const BRTxOutput out
...
         // fee amount after adding a change output
-        feeAmount = _txFee(wallet->feePerKb, BRTransactionVSize(transaction) + TX_OUTPUT_SIZE);
+        if (feeRate > 1000)
+        {
+            feeAmount = _txFee(feeRate, BRTransactionVSize(transaction) + TX_OUTPUT_SIZE);
+        }
+        else
+        {
+            feeAmount = _txFee(wallet->feePerKb, BRTransactionVSize(transaction) + TX_OUTPUT_SIZE);
+        }
```
Comment (bvbfan) [inline src/spv/bitcoin/BRWallet.cpp:1030]: Now feeRate here is adjusted in line 963

File: src/spv/spv_rpc.cpp [hunk 6]
```diff
@@ -871,7 +1090,7 @@ UniValue spv_fundaddress(const JSONRPCRequest& request)
 
     std::string strAddress = request.params[0].get_str();
 
-    return fake_spv->SendBitcoins(pwallet, strAddress, -1);
+    return fake_spv->SendBitcoins(pwallet, strAddress, -1, 1000);
```
Comment (bvbfan) [inline src/spv/spv_rpc.cpp:1093]: We don't need to add 1000, it's just wallet value?

**Review by bvbfan — CHANGES_REQUESTED** [review]

File: src/spv/spv_rpc.cpp (hunk 6 above)
Comment (Bushstar) [reply src/spv/spv_rpc.cpp:1093]: In this circumstance it is never even used!

Comment (Bushstar) [reply src/spv/spv_rpc.cpp:1093]: I remember, feerate was added to SendBitcoins so spv_sendtoaddress can set feerate.

File: src/spv/bitcoin/BRWallet.cpp (hunk 4 above)
Comment (Bushstar) [reply src/spv/bitcoin/BRWallet.cpp]: Changed in https://github.com/DeFiCh/ain/pull/308/commits/a6aae235b22b244bd371a191bb58f0d12c8faab7

Added checking that fee is not less than 1 Sat/byte.

File: src/spv/bitcoin/BRWallet.cpp (hunk 5 above)
Comment (Bushstar) [reply src/spv/bitcoin/BRWallet.cpp]: https://github.com/DeFiCh/ain/pull/308/commits/a6aae235b22b244bd371a191bb58f0d12c8faab7

**Review by bvbfan — APPROVED** [review]

**Review by monstrobishi — APPROVED** [review]

**Pull request merged**